### PR TITLE
Convert any format to any format

### PR DESCRIPTION
**Pull request by binaryberry**
This enables converting any format to any format, except for local transactions which were a special case  (they require a lgsl code, so they are out of the scope of the story).

When reviewing, it'd be good to check I didn't lose any relevant test coverage by deleting the ad-hoc conversion tests and replacing them with the automated/generic version.

For tests to pass, it will require a govuk_content_models gem bump to take into account [these changes](https://github.com/alphagov/govuk_content_models/pull/350).

cc @boffbowsh 

screen shot 2015-11-02 at 16 53 21

[Trello](https://trello.com/c/p7wzEHo0/79-self-serve-changing-a-format-to-almost-any-format-in-mainstream-publisher-medium)

### CODE REVIEW

**Comment by fofr** [comment]
@binaryberry Could you include some screenshots of the UI?

**Comment by binaryberry** [comment]
Sure @fofr 

screen shot 2015-10-30 at 10 03 26
screen shot 2015-10-30 at 10 04 05

**Comment by binaryberry** [comment]
Thanks for the feedback @fofr ! Amended now following this + our conversation in meatspace. Also ordered the editions alphabetically.

File: Gemfile [hunk 1]
```diff
@@ -11,7 +11,7 @@ end
 if ENV['CONTENT_MODELS_DEV']
   gem "govuk_content_models", :path => '../govuk_content_models'
 else
-  gem "govuk_content_models", '31.1.0'
+  gem "govuk_content_models", '31.3.0'
```
Comment (boffbowsh) [inline Gemfile]: If you change this to:
```ruby
gem "govuk_content_models", github: "alphagov/govuk_content_models", branch: "convert-to-any-format"
```
and re-bundle, the tests will be able to run.

File: lib/enhancements/edition.rb [hunk 2]
```diff
@@ -87,4 +87,8 @@ def register_with_panopticon
   def notify_publishing_api
     PublishingAPINotifier.perform_async(self.id.to_s)
   end
+
+  def self.edition_types
+    [AnswerEdition, GuideEdition, TransactionEdition, PlaceEdition, ProgrammeEdition, LicenceEdition, VideoEdition, BusinessSupportEdition, SimpleSmartAnswerEdition, CompletedTransactionEdition, HelpPageEdition, CampaignEdition]
```
Comment (benilovj) [inline lib/enhancements/edition.rb]: is this a manually curated list or is this the list of all edition subtypes? Is `LocalTransactionEdition` deliberately omitted?

It'd be good to not have to change this line if a new format was added or an existing one removed; Rails has [a method](http://apidock.com/rails/v4.2.1/Class/subclasses) for getting subclasses easily, so we could use `Edition.subclasses` in place of `Edition.edition_types`.

File: test/integration/change_edition_type_test.rb [hunk 3]
```diff
@@ -1,261 +1,93 @@
...
-    assert page.has_button?("Create as new Answer edition")
-
-    click_on "Create as new Answer edition"
-
-    assert_text(simple_smart_answer.title)
-    assert_text("New edition created")
-  end
-
-# tests for changing Answer, Guide, Programme into a Transaction
+  edition_types = [:transaction_edition, :place_edition, :programme_edition, :simple_smart_answer_edition, :business_support_edition, :answer_edition, :guide_edition, :licence_edition, :video_edition, :completed_transaction_edition, :help_page_edition, :campaign_edition]
```
Comment (benilovj) [inline test/integration/change_edition_type_test.rb]: Similarly to my previous comment, I'd prefer this to be dynamic on the edition subtypes, not sure what you think @boffbowsh

File: test/integration/change_edition_type_test.rb [hunk 4]
```diff
@@ -1,261 +1,93 @@
 require 'integration_test_helper'
+require 'imminence_areas_test_helper.rb'
```
Comment (benilovj) [inline test/integration/change_edition_type_test.rb]: you don't need the `.rb` here

File: test/integration/change_edition_type_test.rb [hunk 5]
```diff
@@ -1,261 +1,93 @@
...
-    within "div.tabbable" do
-      click_on "Admin"
-    end
+  conversions = edition_types.permutation(2).reject { |pair| pair[0] == pair[1] }
 
-    assert page.has_button?("Create as new Transaction edition")
+  conversions.each do |pair|
 
-    click_on "Create as new Transaction edition"
+    from, to = pair
```
Comment (benilovj) [inline test/integration/change_edition_type_test.rb]: you can avoid this line and unpack `to` and `from` directly:

```ruby
conversions.each do |to, from|
```

File: test/integration/change_edition_type_test.rb [hunk 6]
```diff
@@ -1,261 +1,93 @@
...
-    end
-  end
+      within "div.tabbable" do
+        click_on "Admin"
+      end
 
-  test "should be able to convert an AnswerEdition into a SimpleSmartAnswerEdition" do
-    answer = FactoryGirl.create(:answer_edition, state: 'published', body: "abcde")
-    visit_edition answer
+      select(to.to_s.gsub("_", " ").titleize.gsub(/Edition.*/, 'Edition'), from: 'to')
```
Comment (benilovj) [inline test/integration/change_edition_type_test.rb]: `to.to_s.gsub("_", " ").titleize.gsub(/Edition.*/, 'Edition')` is hard to parse – I'd suggest having a helper method with a clear method name which calculates this.

File: test/integration/change_edition_type_test.rb [hunk 7]
```diff
@@ -1,261 +1,93 @@
...
-    assert_text("New edition created")
-  end
-
-# tests for changing Answer, Guide, Programme into a Transaction
+  edition_types = [:transaction_edition, :place_edition, :programme_edition, :simple_smart_answer_edition, :business_support_edition, :answer_edition, :guide_edition, :licence_edition, :video_edition, :completed_transaction_edition, :help_page_edition, :campaign_edition]
 
-  test "should be able to convert an AnswerEdition into a TransactionEdition" do
-    answer = FactoryGirl.create(:answer_edition, state: 'published')
-    visit_edition answer
+  sample_parts = Set.new([{"title" => "PART !", "body" => "This is some edition version text.", "slug" => "part-one"}, {"title" => "PART !!", "body" => "This is some more edition version text.", "slug" =>  "part-two"}])
```
Comment (benilovj) [inline test/integration/change_edition_type_test.rb]: I think this would be more readable if it were formatted across multiple lines:

```ruby
  sample_parts = Set.new([
    {
      "title" => "PART !",
      "body" => "This is some edition version text.",
      "slug" => "part-one"
    },
    {
      "title" => "PART !!",
      "body" => "This is some more edition version text.",
      "slug" =>  "part-two"
    }
  ])
```

File: test/integration/change_edition_type_test.rb [hunk 8]
```diff
@@ -1,261 +1,93 @@
...
+      edition_whole_body = edition.whole_body.gsub(/\s+/, " ").strip
 
-    assert page.has_content?(answer.title)
-    assert page.has_content?("New edition created")
-    assert page.has_content?("Question 1")
-
-    within :css, '#edition_body_input' do
-      assert page.has_xpath?("//textarea[contains(text(), '#{answer.whole_body}')]"), "Expected to see: #{answer.whole_body}"
+      if edition.respond_to?(:parts)
+        assert(sample_parts.subset?(Set.new(edition.parts.map { |part| part.attributes.slice("title", "body", "slug") })))
```
Comment (benilovj) [inline test/integration/change_edition_type_test.rb]: it's a bit hard to figure out what's going on here. I'd suggest either breaking out some intermediate local variables to help get the intent across, or to introduce a custom assertion method that would help get the intent across.

**Comment by fofr** [comment]
When a guide is switched to something without parts, what happens to the URLs?

File: lib/enhancements/edition.rb (hunk 2 above)
Comment (binaryberry) [inline lib/enhancements/edition.rb]: MUCH better thanks. More readable and adaptable if we add formats. Amended.

File: test/integration/change_edition_type_test.rb (hunk 3 above)
Comment (binaryberry) [inline test/integration/change_edition_type_test.rb]: Makes sense to me, means when a new format is created the conversion tests will update automatically. Paul is on a beach somewhere therefore he agrees. Updated it.

File: test/integration/change_edition_type_test.rb (hunk 4 above)
Comment (binaryberry) [inline test/integration/change_edition_type_test.rb]: thanks, amended!

File: test/integration/change_edition_type_test.rb (hunk 5 above)
Comment (binaryberry) [inline test/integration/change_edition_type_test.rb]: done, thanks!

File: test/integration/change_edition_type_test.rb (hunk 6 above)
Comment (binaryberry) [inline test/integration/change_edition_type_test.rb]: Fair enough - extracted it to select_target_edition method

File: test/integration/change_edition_type_test.rb (hunk 7 above)
Comment (binaryberry) [inline test/integration/change_edition_type_test.rb]: done, thanks!

File: test/integration/change_edition_type_test.rb (hunk 8 above)
Comment (binaryberry) [inline test/integration/change_edition_type_test.rb]: Sure! I've now created the edition_parts method that explains what `Set.new(edition.parts.map { |part| part.attributes.slice("title", "body", "slug") })` is.

**Comment by binaryberry** [comment]
@fofr the URLs of the parts redirect to the URL of the new format, same principle as when you [withdraw]  (https://trello.com/c/BYS3dxxV/118-panopticon-unpublishings-should-redirect-without-keeping-subpath-medium) a document with subpaths.

**Comment by benilovj** [comment]
:+1:

**Comment by jamiecobbett** [comment]
:fireworks:
